[Backtracing][Linux] Add Linux crash handler to the runtime.

al45tair opened the pull request:
This also adds a function to demangle a symbol, and a way for the backtracing code to report warning messages to the same place as the main runtime.

I'd like to rename the `_swift_isThunkFunction()` SPI also, but we can't do that until we've made the changes to the `_Backtracing` library, so we'll do that there instead.

rdar://110261430

al45tair committed:
[Backtracing][Linux] Add Linux crash handler to the runtime.

This also adds a function to demangle a symbol, and a way for the
backtracing code to report warning messages to the same place as
the main runtime.

I'd like to rename the _swift_isThunkFunction() SPI also, but we
can't do that until we've made the changes to the _Backtracing
library, so we'll do that there instead.

rdar://110261430

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing][Linux] Add missing CrashInfo.h header.

This was added to a later PR, but not to this one, though we need
it here.

rdar://110261430

al45tair committed:
[Backtracing][Linux] Disable backtrace-on-crash until enabling PR.

This should have been disabled until #66338.

rdar://110261430

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
If you're wondering where the tests are, I left them until the last PR (#66338) so everything was enabled and they'd actually work.

mikeash reviewed (approved):
A few little comments, but LGTM.

al45tair commented on include/swift/Runtime/CrashInfo.h:
```diff
@@ -0,0 +1,70 @@
+
+struct memserver_resp {
+  uint64_t addr;
+  int64_t  len;
```
I think that file could do with a few more comments in general. I'll add some.

al45tair commented on stdlib/public/runtime/Backtrace.cpp:
```diff
@@ -773,6 +779,54 @@ _swift_backtraceSetupEnvironment()
+do_spawn(void *ptr) {
+  struct spawn_info *pinfo = (struct spawn_info *)ptr;
+
+  /* Ensure that the memory server is always on fd 4 */
```
I'm pondering right now why I chose 4 and not 3. It just has to be something other than 0, 1 and 2, I think. It doesn't _really_ matter what number we choose, as long as both sides of this agree on which one to use. I _could_ have made this an argument to `swift-backtrace` I suppose, but there didn't seem much point.

al45tair commented on stdlib/public/runtime/Backtrace.cpp:
```diff
@@ -773,6 +779,54 @@ _swift_backtraceSetupEnvironment()
+  struct spawn_info info = { path, argv, envp, memserver };
+
+  /* The CLONE_VFORK is *required* because info is on the stack; we don't
+     want to return until *after* the subprocess has called execvpe(). */
```
Yep. We are sprinting with machetes here. We left running with scissors behind some time ago.

al45tair commented on stdlib/public/runtime/CrashHandlerLinux.cpp:
```diff
@@ -0,0 +1,770 @@
+  int fd = memserver_start();
+
+  /* Start the backtracer; this will suspend the process, so there's no need
+     to try to suspend other threads from here. */
```
That comment needs removing, I think. It was that way, once.

al45tair commented on stdlib/public/runtime/CrashHandlerLinux.cpp:
```diff
@@ -0,0 +1,770 @@
+      }
+    }
+
+    // Wait up to 5 seconds for the threads to pause
```
I also need to update the comment here, I think. It still says `SIGUSR1` but it's now using `SIGPROF` because using the former breaks when `libdispatch` is in use.

al45tair replied:
(Yes, this is truly horrid. I don't see any other possibility though. We can definitely get threads into an interruptible wait, and in that case they won't get the signal until that is over, and we don't want to wait indefinitely.)

al45tair commented on stdlib/public/runtime/CrashHandlerLinux.cpp:
```diff
@@ -0,0 +1,770 @@
+    ssize_t ret;
+
+    ret = safe_read(fd, &req, sizeof(req));
+    if (ret != sizeof(req))
```
No. According to its man page, `process_vm_readv()` will always read a complete `iovec` element, never a partial one, and our `memcpy()` always does the same.

al45tair replied:
Good point. I think that _is_ possible, actually. I will update the code and check the other side too.

al45tair commented on stdlib/public/runtime/CrashHandlerLinux.cpp:
```diff
@@ -0,0 +1,770 @@
+      resp.len = ret;
+
+      ret = safe_write(fd, &resp, sizeof(resp));
+      if (ret != sizeof(resp))
```
Same answer.

MaxDesiatov commented on include/swift/Runtime/Backtrace.h:
```diff
@@ -50,7 +57,11 @@ typedef int ErrorCode;
 SWIFT_RUNTIME_STDLIB_INTERNAL ErrorCode _swift_installCrashHandler();
 
+#ifdef __linux__
+SWIFT_RUNTIME_STDLIB_INTERNAL bool _swift_spawnBacktracer(const ArgChar * const *argv, int memserver_fd);
```
Moving my feedback from the other PR: would this benefit from a doc comment that clarifies the use of each parameter and the return pointer as well? Even if it's already in the `.cpp` file, IUUC editors with support for displaying documentation on hovered symbol load it from `.h` files.

al45tair committed:
[Backtracing][Linux] Tidy a few things up after review.

Mike and Max made various helpful suggestions, so I've added and updated
various comments and amended the code to cope with partial reads and
writes.

rdar://110261430

al45tair commented on stdlib/public/runtime/Backtrace.cpp:
```diff
@@ -796,13 +850,109 @@ _swift_isThunkFunction(const char *mangledName) {
+        return result;
+      }
+
+      size_t toCopy = std::min(bufferSize - 1, resultLen - 1);
```
Fixed.

al45tair commented on stdlib/public/runtime/CrashHandlerLinux.cpp:
```diff
@@ -0,0 +1,770 @@
+    if (ss.ss_sp == MAP_FAILED)
+      return errno;
+
+    if (sigaltstack(&ss, NULL) < 0)
```
I don't think I'm too worried about that. The other error paths in the function leak it also; it's relatively small, and I don't really expect those error paths to trigger on a system in a sane state.

al45tair commented on stdlib/public/runtime/CrashHandlerLinux.cpp:
```diff
@@ -0,0 +1,770 @@
+
+    // See if a signal handler for this signal is already installed
+    if (sigaction(signalsToHandle[n], NULL, &osa) < 0)
+      return errno;
```
I _think_ all of the failure modes for `sigaction()` are programming errors, so that error path shouldn't ever happen in practice.

al45tair committed:
[Backtracing][Linux] Move doc comment to header.

Moved the comment for `_swift_backtrace_demangle` into the header file
instead of it being in the implementation.

rdar://110261430

MaxDesiatov reviewed (approved):
Thanks!
